File: vscode-extension/src/apiClient.ts
// vscode-extension/src/apiClient.ts
// -----------------------------------------------------------------------------
// Thin client for the backend's existing REST contract — see
// api/routes/analysis.py. No new backend endpoints were needed for this
// extension; POST /analyse already accepts a raw diff_text and returns a
// request_id to poll, which is exactly what an editor-loop client needs.
//
// Uses the global `fetch` (Node 18+, available in the VS Code extension host
// — no extra HTTP dependency needed).

import { PathReviewConfig } from './pathReviewConfig';
import { ModelOverride } from './settings';

export interface AnalyzeOptions {
  backendUrl: string;
  apiKey: string;
  repoUrl: string;
  sourceRef: string;
  targetRef?: string;
  diffText: string;
  selectedAgents: string[] | null;
  /** Free-text prioritization guidance — scanned client-side (promptGuard.ts)
   * before this is ever called, and re-validated authoritatively server-side
   * (governance/prompt_guard.py) regardless. */
  userInstructions?: string;
  /** Parsed .gto.yaml (pathReviewConfig.ts) — re-validated/re-scanned
   * server-side (core/orchestrator.py) before any of its free text reaches
   * an LLM prompt; this client-side parse is not itself a trust boundary. */
  pathReviewConfig?: PathReviewConfig;
  /** Optional LLM override (settings.ts::getModelOverride) — matches the web
   * app's `llm_config` field exactly (api/routes/analysis.py's
   * AnalyseRequest.llm_config). Undefined = don't send it, use whatever the
   * backend is configured with (today's behavior, unchanged). */
  modelOverride?: ModelOverride;
}

export interface SubmitResponse {
  request_id: string;
  status: 'queued' | 'no_diff' | string;
  message?: string;
}

export interface StatusResponse {
  request_id: string;
  status: 'queued' | 'running' | 'done' | 'unknown' | string;
  queue_position?: number;
  queue_total?: number;
}

export interface CorrelatedIssue {
  title: string;
  file_path: string;
  line: number;
  severity: string;
  confidence: string;
  score: number;
  agents: string[];
  categories: string[];
  descriptions: string[];
}

/** A concrete, copy-pasteable suggested fix — deterministic (not LLM), only
 * populated when the `remediation` agent runs (currently: Thorough preset
 * only). `diff` embeds the source line number as "@@ line N @@" since CodeFix
 * itself doesn't carry a line field — see agents/fix_generator.py. */
export interface CodeFix {
  title: string;
  file_path: string;
  category: string;
  severity: string;
  before: string;
  after: string;
  diff: string;
  explanation: string;
  confidence: string;
}

/** One missing/at-risk test scenario found by the `qa_scenarios` agent
 * (Thorough preset only). `test_skeleton` is a real, language-aware
 * Arrange/Act/Assert stub — generated deterministically from the actual
 * changed symbol, not LLM prose — see agents/qa_scenarios_agent.py. */
export interface QAScenario {
  id: string;
  title: string;
  type: string;
  priority: string;
  description: string;
  steps: string[];
  expected_result: string;
  affected_files: string[];
  automation_hint: string;
  preconditions: string[];
  acceptance_criteria: string[];
  test_skeleton: string;
  test_skeleton_filename: string;
}

/** An LLM-narrated (not call-graph-grounded) sequence diagram for a complex
 * change — see agents/remediation_agent.py::_maybe_generate_diagram. Only
 * generated on the Thorough preset, and only when the change has real
 * reference_impact data AND medium+ risk — most reports won't have one.
 * confidence is hardcoded "low" server-side, never LLM-settable — same
 * "review before relying on it" framing as an AI-suggested code fix. */
export interface MermaidDiagram {
  diagram_type: string;
  mermaid_source: string;
  confidence: string;
  note: string;
}

export interface QAScenariosResult {
  scenarios: QAScenario[];
  total_scenarios: number;
  critical_count: number;
  high_count: number;
  coverage_areas: string[];
  summary: string;
}

/** Display-only summary of what .gto.yaml did for this run — see
 * core.models.PathReviewSummary. Absent/undefined when no .gto.yaml (or an
 * empty one) was in play, the common case. */
export interface PathReviewSummary {
  agents_excluded: string[];
  steering_applied: boolean;
}

export interface AnalysisReport {
  request_id: string;
  repo_url: string;
  gate_decision: string;
  final_risk: string;
  risk?: { risk_score?: number; rationale?: string };
  remediation?: { code_fixes?: CodeFix[]; fix_suggestions?: string[]; diagrams?: MermaidDiagram[] };
  qa_scenarios?: QAScenariosResult;
  path_review_summary?: PathReviewSummary | null;
  files_changed: number;
  files_changed_list: string[];
  top_issues: CorrelatedIssue[];
  duration_s: number;
  errors: string[];
}

export class ApiError extends Error {
  constructor(message: string, readonly status?: number) {
    super(message);
  }
}

/** Node's global fetch() collapses every network-level failure (DNS,
 * connection refused, timeout, TLS) into a generic "fetch failed" — the
 * actual reason lives on `err.cause`, which a plain `.message` read drops
 * silently. Walks the cause chain so the real reason (e.g. "connect
 * ECONNREFUSED 10.0.0.5:8080") reaches the user instead of a dead end. Only
 * matters for errors that never got an HTTP response — ApiError (a real
 * response came back) already carries a useful, backend-provided message. */
export function describeError(e: unknown): string {
  if (!(e instanceof Error)) return String(e);
  const parts = [e.message];
  let cause: unknown = (e as { cause?: unknown }).cause;
  let depth = 0;
  while (cause && depth < 5) {
    if (cause instanceof Error) {
      parts.push(cause.message);
      cause = (cause as { cause?: unknown }).cause;
    } else {
      parts.push(String(cause));
      cause = undefined;
    }
    depth++;
  }
  return parts.filter((p, i) => parts.indexOf(p) === i).join(' — caused by: ');
}

function headers(apiKey: string): Record<string, string> {
  return { 'Content-Type': 'application/json', 'X-API-Key': apiKey };
}

async function parseErrorBody(resp: Response): Promise<string> {
  try {
    const body = (await resp.json()) as { detail?: string | { msg?: string }[] };
    const d = body?.detail;
    if (typeof d === 'string') return d;
    if (Array.isArray(d)) return d.map((x) => x.msg).filter(Boolean).join('; ');
  } catch {
    /* body wasn't JSON — fall through to the generic message below */
  }
  return `Backend error ${resp.status}`;
}

/** A named model choice the backend admin configured (config/settings.py's
 * MODEL_PRESETS) — secret-free (no base_url/api_key), see
 * settings.ts::getSelectedModelPreset for how a picked preset becomes a
 * ModelOverride. */
export interface ModelPreset {
  name: string;
  label: string;
  provider: string;
  model: string;
}

export async function fetchModelPresets(backendUrl: string, apiKey: string): Promise<ModelPreset[]> {
  const resp = await fetch(`${backendUrl}/api/v1/model-presets`, { headers: headers(apiKey) });
  if (resp.status === 401) throw new ApiError('Unauthorized — check your API key (GTO: Set API Key).', 401);
  if (!resp.ok) throw new ApiError(await parseErrorBody(resp), resp.status);
  const body = (await resp.json()) as { presets?: ModelPreset[] };
  return body.presets ?? [];
}

/** One past analysis similar to the current one — file-overlap + summary-keyword
 * similarity, see api/routes/insights.py::similar_prs. Same web-app feature
 * already shown in the frontend's ResultsView, now also in the panel. */
export interface SimilarPR {
  request_id: string;
  repo: string;
  pr_title: string;
  pr_number: number;
  source_ref: string;
  risk_score: number;
  gate: string;
  similarity: number;
  shared_files: string[];
  elapsed: string;
}

/** Best-effort — a nice-to-have context panel, never worth surfacing an error
 * for. Callers should treat a thrown/empty result as "just hide the section." */
export async function fetchSimilarPRs(backendUrl: string, apiKey: string, requestId: string, topK = 5): Promise<SimilarPR[]> {
  const resp = await fetch(`${backendUrl}/api/v1/insights/similar/${requestId}?top_k=${topK}`, { headers: headers(apiKey) });
  if (!resp.ok) return [];
  const body = (await resp.json()) as { similar?: SimilarPR[] };
  return body.similar ?? [];
}

/** Records a reviewer verdict on one finding — the same feedback loop the web
 * app's ResultsView already exposes (⚐/✓ controls). Aggregated over ≥3
 * false_positive verdicts for the same (agent, category, repo),
 * governance/suppression.py auto-suppresses that pattern on future runs of
 * this repo — always visibly, via report.suppressed_notes, never silently. */
/** Posts findings as PR comments (one grouped comment per file + an overall
 * summary) — reuses the existing POST /report/{id}/comment-pr endpoint
 * unchanged. Deliberately omits token/base_url/workspace so the backend
 * falls back to its own shared bot credential (the same identity webhook-
 * triggered comments already post under) — no personal token needed. */
export async function postFindingsToPR(
  backendUrl: string,
  apiKey: string,
  requestId: string,
  opts: { repoSlug: string; prId: string; provider: string }
): Promise<{ ok: boolean; comments_posted: number; files_commented: number }> {
  const resp = await fetch(`${backendUrl}/api/v1/report/${requestId}/comment-pr`, {
    method: 'POST',
    headers: headers(apiKey),
    body: JSON.stringify({ provider: opts.provider, repo_slug: opts.repoSlug, pr_id: opts.prId, inline: true }),
  });
  if (resp.status === 401) throw new ApiError('Unauthorized — check your API key (GTO: Set API Key).', 401);
  if (!resp.ok) throw new ApiError(await parseErrorBody(resp), resp.status);
  return (await resp.json()) as { ok: boolean; comments_posted: number; files_commented: number };
}

/** Reviewer sign-off — approval only, never merges. Requires the reviewer's
 * OWN token (no server-side fallback, unlike postFindingsToPR) so the
 * approval shows as them, not the shared bot — see settings.ts's
 * getBitbucketToken doc comment for why. Distinct endpoint from gate
 * override: this never changes GTO's own gate decision. */
export async function approvePR(
  backendUrl: string,
  apiKey: string,
  requestId: string,
  opts: { provider: string; token: string; repoSlug: string; prId: string }
): Promise<{ status: string; pr_action: { ok: boolean; errors?: string[] } }> {
  const resp = await fetch(`${backendUrl}/api/v1/gate/${requestId}/approve-pr`, {
    method: 'POST',
    headers: headers(apiKey),
    body: JSON.stringify({ provider: opts.provider, token: opts.token, repo_slug: opts.repoSlug, pr_id: opts.prId }),
  });
  if (resp.status === 401) throw new ApiError('Unauthorized — check your API key (GTO: Set API Key).', 401);
  if (resp.status === 403) throw new ApiError('Your API key does not have PR-approval permission.', 403);
  if (!resp.ok) throw new ApiError(await parseErrorBody(resp), resp.status);
  return (await resp.json()) as { status: string; pr_action: { ok: boolean; errors?: string[] } };
}

/** "Explain this finding" — a fixed, never-user-typed question, answered by
 * the same guardrailed Q&A engine already used for PR chat replies. */
export async function explainFinding(
  backendUrl: string,
  apiKey: string,
  requestId: string,
  body: { agent: string; category?: string; file_path?: string; title?: string }
): Promise<string> {
  const resp = await fetch(`${backendUrl}/api/v1/report/${requestId}/explain-finding`, {
    method: 'POST',
    headers: headers(apiKey),
    body: JSON.stringify(body),
  });
  if (resp.status === 401) throw new ApiError('Unauthorized — check your API key (GTO: Set API Key).', 401);
  if (!resp.ok) {
    const detail = await parseErrorBody(resp);
    // A bare "Not Found" (FastAPI's unknown-route body) means this backend
    // build predates the /explain-finding endpoint — distinct from
    // "Report '…' not found." which is a real, expired report.
    if (resp.status === 404 && /^not found\.?$/i.test(detail.trim())) {
      throw new ApiError(
        'This GTO backend is too old to support Explain — ask your admin to update it to the latest build.',
        404,
      );
    }
    throw new ApiError(detail, resp.status);
  }
  const data = (await resp.json()) as { answer?: string };
  return data.answer ?? '';
}

export async function submitFindingFeedback(
  backendUrl: string,
  apiKey: string,
  requestId: string,
  body: { agent: string; category?: string; file_path?: string; verdict: string; note?: string }
): Promise<void> {
  const resp = await fetch(`${backendUrl}/api/v1/report/${requestId}/feedback`, {
    method: 'POST',
    headers: headers(apiKey),
    body: JSON.stringify(body),
  });
  if (resp.status === 401) throw new ApiError('Unauthorized — check your API key (GTO: Set API Key).', 401);
  if (!resp.ok) throw new ApiError(await parseErrorBody(resp), resp.status);
}

export async function submitAnalysis(opts: AnalyzeOptions): Promise<SubmitResponse> {
  const resp = await fetch(`${opts.backendUrl}/api/v1/analyse`, {
    method: 'POST',
    headers: headers(opts.apiKey),
    body: JSON.stringify({
      repo_url: opts.repoUrl,
      source_ref: opts.sourceRef,
      target_ref: opts.targetRef ?? 'HEAD',
      change_type: 'branch_diff',
      diff_text: opts.diffText,
      selected_agents: opts.selectedAgents,
      user_instructions: opts.userInstructions ?? '',
      path_review_config: opts.pathReviewConfig ?? null,
      ...(opts.modelOverride ? { llm_config: opts.modelOverride } : {}),
    }),
  });
  if (resp.status === 401) throw new ApiError('Unauthorized — check your API key (GTO: Set API Key).', 401);
  if (!resp.ok) throw new ApiError(await parseErrorBody(resp), resp.status);
  return (await resp.json()) as SubmitResponse;
}

export async function getStatus(backendUrl: string, apiKey: string, requestId: string): Promise<StatusResponse> {
  const resp = await fetch(`${backendUrl}/api/v1/status/${requestId}`, { headers: headers(apiKey) });
  if (!resp.ok) throw new ApiError(await parseErrorBody(resp), resp.status);
  return (await resp.json()) as StatusResponse;
}

export async function getReport(backendUrl: string, apiKey: string, requestId: string): Promise<AnalysisReport> {
  const resp = await fetch(`${backendUrl}/api/v1/report/${requestId}?fmt=full`, { headers: headers(apiKey) });
  if (!resp.ok) throw new ApiError(await parseErrorBody(resp), resp.status);
  return (await resp.json()) as AnalysisReport;
}

/** Polls /status until done, then fetches the full report. Cancelable via `token`. */
export async function runAnalysisToCompletion(
  opts: AnalyzeOptions,
  token: { cancelled: boolean },
  onStatus?: (s: StatusResponse) => void,
  pollMs = 2000
): Promise<AnalysisReport | null> {
  const submitted = await submitAnalysis(opts);
  if (submitted.status === 'no_diff') {
    return null;
  }
  const requestId = submitted.request_id;
  while (!token.cancelled) {
    await new Promise((r) => setTimeout(r, pollMs));
    const s = await getStatus(opts.backendUrl, opts.apiKey, requestId);
    onStatus?.(s);
    if (s.status === 'done') {
      return getReport(opts.backendUrl, opts.apiKey, requestId);
    }
    if (s.status === 'unknown') {
      throw new ApiError('Analysis vanished server-side (restarted backend?) — please retry.');
    }
  }
  return null;
}

File: intellij-plugin/src/main/kotlin/com/uob/gto/api/ApiClient.kt
package com.uob.gto.api

import kotlinx.coroutines.delay
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.put
import kotlinx.serialization.json.putJsonArray
import kotlinx.serialization.json.putJsonObject
import java.net.URI
import java.net.http.HttpClient
import java.net.http.HttpRequest
import java.net.http.HttpResponse
import java.time.Duration

/**
 * Thin client for the backend's existing REST contract — mirrors
 * vscode-extension/src/apiClient.ts exactly (same endpoints, same request
 * shape), so a review submitted from IntelliJ is indistinguishable
 * server-side from one submitted through VS Code or the web app.
 *
 * Deliberately does NOT deserialize AnalysisReport into a Kotlin data class:
 * the report is rendered entirely by the shared JS bundle
 * (shared/report-view, loaded into a JCEF browser — see toolwindow/), which
 * consumes the raw JSON directly via JSON.parse in the browser. Kotlin only
 * needs to read a handful of top-level fields (gate_decision, request_id) —
 * see [AnalysisReportView] — not the whole ~40-field contract.
 */
class ApiException(message: String, val status: Int = 0) : Exception(message)

data class AnalyzeOptions(
    val backendUrl: String,
    val apiKey: String,
    val repoUrl: String,
    val sourceRef: String,
    val targetRef: String = "HEAD",
    val diffText: String,
    val selectedAgents: List<String>?,
    val userInstructions: String = "",
    val pathReviewConfig: JsonObject? = null,
    val modelOverride: JsonObject? = null,
    val metadata: JsonObject? = null,
)

data class SubmitResponse(val requestId: String, val status: String, val message: String?)
data class StatusResponse(val requestId: String, val status: String, val queuePosition: Int?, val queueTotal: Int?)

/** Thin read-only view over the raw report JSON for the handful of fields
 * Kotlin-side logic (notifications, inspections, "Show Last Result") needs
 * to inspect without round-tripping through a full data class. */
class AnalysisReportView(val raw: JsonObject) {
    val requestId: String get() = raw["request_id"]?.jsonPrimitive?.content ?: ""
    val gateDecision: String get() = raw["gate_decision"]?.jsonPrimitive?.content ?: ""
    val topIssuesCount: Int get() = raw["top_issues"]?.jsonArray?.size ?: 0
    val filesChanged: Int get() = raw["files_changed"]?.jsonPrimitive?.content?.toIntOrNull() ?: 0
    /** Raw JSON text — handed verbatim to the JCEF browser's JSON.parse(). */
    val json: String get() = raw.toString()
}

class ApiClient(private val backendUrl: String, private val apiKey: String) {

    // Pinned to HTTP/1.1: the JDK client's default HTTP/2 preference sends an
    // "Upgrade: h2c" preface on the first cleartext (http://) request, which
    // uvicorn's h11 parser can't handle and rejects with "Invalid HTTP
    // request received." — most visibly on POST /analyse (a body-bearing
    // request), while bodyless GETs could slip through.
    private val http: HttpClient = HttpClient.newBuilder()
        .connectTimeout(Duration.ofSeconds(15))
        .version(HttpClient.Version.HTTP_1_1)
        .build()

    private fun request(path: String, method: String = "GET", body: String? = null): HttpRequest.Builder {
        val builder = HttpRequest.newBuilder()
            .uri(URI.create("$backendUrl$path"))
            .header("Content-Type", "application/json")
            .header("X-API-Key", apiKey)
            .timeout(Duration.ofSeconds(60))
        return when (method) {
            "POST" -> builder.POST(HttpRequest.BodyPublishers.ofString(body ?: "{}"))
            else -> builder.GET()
        }
    }

    private fun checkOk(resp: HttpResponse<String>) {
        if (resp.statusCode() == 401) {
            throw ApiException("Unauthorized — check your API key (GTO Review settings).", 401)
        }
        if (resp.statusCode() !in 200..299) {
            val detail = try {
                Json.parseToJsonElement(resp.body()).jsonObject["detail"]?.jsonPrimitive?.content
            } catch (e: Exception) {
                null
            }
            throw ApiException(detail ?: resp.body().take(300).ifBlank { "HTTP ${resp.statusCode()}" }, resp.statusCode())
        }
    }

    fun submitAnalysis(opts: AnalyzeOptions): SubmitResponse {
        val body = buildJsonObject {
            put("repo_url", opts.repoUrl)
            put("source_ref", opts.sourceRef)
            put("target_ref", opts.targetRef)
            put("change_type", "branch_diff")
            put("diff_text", opts.diffText)
            if (opts.selectedAgents != null) {
                putJsonArray("selected_agents") { opts.selectedAgents.forEach { add(JsonPrimitive(it)) } }
            } else {
                put("selected_agents", JsonPrimitive(null as String?))
            }
            put("user_instructions", opts.userInstructions)
            put("path_review_config", opts.pathReviewConfig ?: JsonPrimitive(null as String?))
            if (opts.modelOverride != null) put("llm_config", opts.modelOverride)
            if (opts.metadata != null && opts.metadata.isNotEmpty()) put("metadata", opts.metadata)
        }
        val resp = http.send(request("/api/v1/analyse", "POST", body.toString()).build(), HttpResponse.BodyHandlers.ofString())
        checkOk(resp)
        val json = Json.parseToJsonElement(resp.body()).jsonObject
        return SubmitResponse(
            requestId = json["request_id"]?.jsonPrimitive?.content ?: "",
            status = json["status"]?.jsonPrimitive?.content ?: "unknown",
            message = json["message"]?.jsonPrimitive?.contentOrNull,
        )
    }

    fun getStatus(requestId: String): StatusResponse {
        val resp = http.send(request("/api/v1/status/$requestId").build(), HttpResponse.BodyHandlers.ofString())
        checkOk(resp)
        val json = Json.parseToJsonElement(resp.body()).jsonObject
        return StatusResponse(
            requestId = json["request_id"]?.jsonPrimitive?.content ?: requestId,
            status = json["status"]?.jsonPrimitive?.content ?: "unknown",
            queuePosition = json["queue_position"]?.jsonPrimitive?.content?.toIntOrNull(),
            queueTotal = json["queue_total"]?.jsonPrimitive?.content?.toIntOrNull(),
        )
    }

    fun getReport(requestId: String): AnalysisReportView {
        val resp = http.send(request("/api/v1/report/$requestId?fmt=full").build(), HttpResponse.BodyHandlers.ofString())
        checkOk(resp)
        return AnalysisReportView(Json.parseToJsonElement(resp.body()).jsonObject)
    }

    /** "Explain this finding" — a fixed, never-user-typed question, answered
     * by the same guardrailed Q&A engine used for PR chat replies.
     * [modelOverride], when given, is the same llm_config shape AnalyzeOptions
     * sends — lets Explain reuse whichever provider/model the analysis itself
     * used instead of the backend's global default model. */
    fun explainFinding(
        requestId: String, agent: String, category: String?, filePath: String?, title: String?,
        modelOverride: JsonObject? = null,
    ): String {
        val body = buildJsonObject {
            put("agent", agent)
            if (category != null) put("category", category)
            if (filePath != null) put("file_path", filePath)
            if (title != null) put("title", title)
            if (modelOverride != null) put("llm_config", modelOverride)
        }
        val resp = http.send(request("/api/v1/report/$requestId/explain-finding", "POST", body.toString()).build(), HttpResponse.BodyHandlers.ofString())
        if (resp.statusCode() == 404) {
            val detail = try { Json.parseToJsonElement(resp.body()).jsonObject["detail"]?.jsonPrimitive?.content } catch (e: Exception) { null }
            if (detail != null && detail.trim().lowercase().matches(Regex("not found\\.?"))) {
                throw ApiException("This GTO backend is too old to support Explain — ask your admin to update it to the latest build.", 404)
            }
        }
        checkOk(resp)
        return Json.parseToJsonElement(resp.body()).jsonObject["answer"]?.jsonPrimitive?.content ?: ""
    }

    /** Records a reviewer verdict on one finding — the same feedback loop the
     * web app's ResultsView already exposes. Aggregated over ≥3
     * false_positive verdicts for the same (agent, category, repo),
     * governance/suppression.py auto-suppresses that pattern on future runs. */
    fun submitFindingFeedback(requestId: String, agent: String, category: String?, filePath: String?, verdict: String, note: String? = null) {
        val body = buildJsonObject {
            put("agent", agent)
            if (category != null) put("category", category)
            if (filePath != null) put("file_path", filePath)
            put("verdict", verdict)
            if (note != null) put("note", note)
        }
        val resp = http.send(request("/api/v1/report/$requestId/feedback", "POST", body.toString()).build(), HttpResponse.BodyHandlers.ofString())
        checkOk(resp)
    }

    data class SimilarPR(
        val requestId: String, val repo: String, val prTitle: String, val prNumber: Int,
        val sourceRef: String, val riskScore: Int, val gate: String, val similarity: Double,
        val sharedFiles: List<String>, val elapsed: String,
    )

    /** Best-effort — a nice-to-have context panel, never worth surfacing an
     * error for. Returns empty on any failure, same as apiClient.ts. */
    fun fetchSimilarPRs(requestId: String, topK: Int = 5): List<SimilarPR> {
        return try {
            val resp = http.send(request("/api/v1/insights/similar/$requestId?top_k=$topK").build(), HttpResponse.BodyHandlers.ofString())
            if (resp.statusCode() !in 200..299) return emptyList()
            val body = Json.parseToJsonElement(resp.body()).jsonObject
            val arr = body["similar"]?.jsonArray ?: return emptyList()
            arr.map {
                val o = it.jsonObject
                SimilarPR(
                    requestId = o["request_id"]?.jsonPrimitive?.content ?: "",
                    repo = o["repo"]?.jsonPrimitive?.content ?: "",
                    prTitle = o["pr_title"]?.jsonPrimitive?.content ?: "",
                    prNumber = o["pr_number"]?.jsonPrimitive?.content?.toIntOrNull() ?: 0,
                    sourceRef = o["source_ref"]?.jsonPrimitive?.content ?: "",
                    riskScore = o["risk_score"]?.jsonPrimitive?.content?.toIntOrNull() ?: 0,
                    gate = o["gate"]?.jsonPrimitive?.content ?: "",
                    similarity = o["similarity"]?.jsonPrimitive?.content?.toDoubleOrNull() ?: 0.0,
                    sharedFiles = o["shared_files"]?.jsonArray?.map { f -> f.jsonPrimitive.content } ?: emptyList(),
                    elapsed = o["elapsed"]?.jsonPrimitive?.content ?: "",
                )
            }
        } catch (e: Exception) {
            emptyList()
        }
    }

    data class PostToPrResult(val ok: Boolean, val commentsPosted: Int, val filesCommented: Int)

    /** Posts findings as PR comments — reuses the shared bot credential
     * server-side, no personal token needed (unlike [approvePr]). */
    fun postFindingsToPr(requestId: String, provider: String, repoSlug: String, prId: String): PostToPrResult {
        val body = buildJsonObject {
            put("provider", provider)
            put("repo_slug", repoSlug)
            put("pr_id", prId)
            put("inline", true)
        }
        val resp = http.send(request("/api/v1/report/$requestId/comment-pr", "POST", body.toString()).build(), HttpResponse.BodyHandlers.ofString())
        checkOk(resp)
        val o = Json.parseToJsonElement(resp.body()).jsonObject
        return PostToPrResult(
            ok = (o["ok"] as? kotlinx.serialization.json.JsonPrimitive)?.content?.toBoolean() ?: false,
            commentsPosted = o["comments_posted"]?.jsonPrimitive?.content?.toIntOrNull() ?: 0,
            filesCommented = o["files_commented"]?.jsonPrimitive?.content?.toIntOrNull() ?: 0,
        )
    }

    data class ApprovePrResult(val status: String, val ok: Boolean, val errors: List<String>)

    /** Reviewer sign-off — approval only, never merges. Requires the
     * reviewer's OWN token (no server-side fallback) so it shows as them,
     * not the shared bot. */
    fun approvePr(requestId: String, provider: String, token: String, repoSlug: String, prId: String): ApprovePrResult {
        val body = buildJsonObject {
            put("provider", provider)
            put("token", token)
            put("repo_slug", repoSlug)
            put("pr_id", prId)
        }
        val resp = http.send(request("/api/v1/gate/$requestId/approve-pr", "POST", body.toString()).build(), HttpResponse.BodyHandlers.ofString())
        if (resp.statusCode() == 403) throw ApiException("Your API key does not have PR-approval permission.", 403)
        checkOk(resp)
        val o = Json.parseToJsonElement(resp.body()).jsonObject
        val prAction = o["pr_action"]?.jsonObject
        return ApprovePrResult(
            status = o["status"]?.jsonPrimitive?.content ?: "",
            ok = (prAction?.get("ok") as? kotlinx.serialization.json.JsonPrimitive)?.content?.toBoolean() ?: false,
            errors = prAction?.get("errors")?.jsonArray?.map { it.jsonPrimitive.content } ?: emptyList(),
        )
    }

    data class ModelPreset(val name: String, val label: String, val provider: String, val model: String)

    fun fetchModelPresets(): List<ModelPreset> {
        val resp = http.send(request("/api/v1/model-presets").build(), HttpResponse.BodyHandlers.ofString())
        checkOk(resp)
        val arr = Json.parseToJsonElement(resp.body()).jsonObject["presets"]?.jsonArray ?: return emptyList()
        return arr.map {
            val o = it.jsonObject
            ModelPreset(
                name = o["name"]?.jsonPrimitive?.content ?: "",
                label = o["label"]?.jsonPrimitive?.content ?: "",
                provider = o["provider"]?.jsonPrimitive?.content ?: "",
                model = o["model"]?.jsonPrimitive?.content ?: "",
            )
        }
    }

    /** Polls /status until done, then fetches the full report. `cancelled`
     * is checked between polls (mirrors apiClient.ts::runAnalysisToCompletion's
     * cancellation token, driven by IntelliJ's ProgressIndicator instead of
     * VS Code's CancellationToken). */
    suspend fun runAnalysisToCompletion(
        opts: AnalyzeOptions,
        isCancelled: () -> Boolean,
        onStatus: (StatusResponse) -> Unit = {},
        pollMs: Long = 2000,
    ): AnalysisReportView? {
        val submitted = submitAnalysis(opts)
        if (submitted.status == "no_diff") return null
        var status = getStatus(submitted.requestId)
        onStatus(status)
        while (status.status != "done" && !isCancelled()) {
            delay(pollMs)
            status = getStatus(submitted.requestId)
            onStatus(status)
        }
        if (isCancelled()) return null
        return getReport(submitted.requestId)
    }
}
